File: AspNetMvcUnleashed/0672329980_source/Chapter10/CS/MvcApplication1/MvcApplication1/CustomActionFilters/FeaturedProductActionFilter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web.Mvc;
using MvcApplication1.Models;

namespace MvcApplication1.CustomActionFilters
{
    public class FeaturedProductAttribute: ActionFilterAttribute
    {
        private ProductsDBEntities _entities = new ProductsDBEntities(); 

        public override void OnResultExecuting(ResultExecutingContext filterContext)
        {
            var viewData = filterContext.Controller.ViewData;
            viewData["featured"] = GetRandomProducts();
        }

        private IList<Product> GetRandomProducts()
        {
            var rnd = new Random();
            var allProducts = _entities.ProductSet.ToList();
            var featuredProducts = new List<Product>();
            for (int i = 0; i < 3; i++)
            {
                var product = allProducts[rnd.Next(allProducts.Count)];
                allProducts.Remove(product);
                featuredProducts.Add(product);
            }
            return featuredProducts;
        }

    }
}

File: AspNetMvcUnleashed/0672329980_source/Chapter10/VB/MvcApplication1/MvcApplication1/CustomActionFilters/FeaturedProductActionFilter.vb
Public Class FeaturedProductAttribute
    Inherits ActionFilterAttribute

    Private _entities As New ProductsDBEntities()

    Public Overrides Sub OnResultExecuting(ByVal filterContext As ResultExecutingContext)
        Dim viewData = filterContext.Controller.ViewData
        viewData("featured") = GetRandomProducts()
    End Sub

    Private Function GetRandomProducts() As IList(Of Product)
        Dim rnd = New Random()
        Dim allProducts = _entities.ProductSet.ToList()
        Dim featuredProducts = New List(Of Product)()
        For i As Integer = 0 To 2
            Dim product = allProducts(rnd.Next(allProducts.Count))
            allProducts.Remove(product)
            featuredProducts.Add(product)
        Next
        Return featuredProducts
    End Function

End Class
